src/frontend/src/handler/create_table.rs

```diff
@@ -178,22 +162,137 @@ pub fn bind_sql_columns(
             name: name.real_value(),
             field_descs,
             type_name: "".to_string(),
+            generated_column: None,
         });
     }
 
-    Ok((column_descs, pk_column_id))
+    Ok(column_descs)
+}
+
+fn check_generated_column_constraints(
+    column_name: &String,
+    expr: &ExprImpl,
+    column_catalogs: &[ColumnCatalog],
+    generated_column_names: &[String],
+) -> Result<()> {
+    let input_refs = expr.collect_input_refs(column_catalogs.len());
+    for idx in input_refs.ones() {
+        let referred_generated_column = &column_catalogs[idx].column_desc.name;
+        if generated_column_names
+            .iter()
+            .any(|c| c == referred_generated_column)
+        {
+            return Err(ErrorCode::BindError(
+                format!("Generated can not reference another generated column, but here generated column \"{}\" referenced another generated column \"{}\"", column_name, referred_generated_column),
+            )
+            .into());
+        }
+    }
+    Ok(())
 }
 
-/// Binds table constraints given the binding results from column definitions.
+/// Binds constraints that can be only specified in column definitions.
+pub fn bind_sql_column_constraints(
+    session: &SessionImpl,
+    table_name: String,
+    column_catalogs: &mut [ColumnCatalog],
+    columns: Vec<ColumnDef>,
+) -> Result<()> {
+    let generated_column_names = {
+        let mut names = vec![];
+        for column in &columns {
+            for option_def in &column.options {
+                if let ColumnOption::GeneratedColumns(_) = option_def.option {
+                    names.push(column.name.real_value());
+                    break;
+                }
+            }
+        }
+        names
+    };
+
+    let mut binder = Binder::new_for_stream(session);
+    binder.bind_columns_to_context(table_name.clone(), column_catalogs.to_vec())?;
+    for column in columns {
+        for option_def in column.options {
+            match option_def.option {
+                ColumnOption::GeneratedColumns(expr) => {
+                    let idx = binder
+                        .get_column_binding_index(table_name.clone(), &column.name.real_value())?;
+                    let expr_impl = binder.bind_expr(expr)?;
+
+                    check_generated_column_constraints(
+                        &column.name.real_value(),
+                        &expr_impl,
+                        column_catalogs,
+                        &generated_column_names,
+                    )?;
+
+                    column_catalogs[idx].column_desc.generated_column = Some(GeneratedColumnDesc {
+                        expr: Some(expr_impl.to_expr_proto()),
+                    });
+                }
+                ColumnOption::Unique { is_primary: true } => {
+                    // Bind primary key in `bind_sql_table_column_constraints`
```

Comment:
What if a generated column is marked as (part of) the primary key? I tend to forbid this.

Reply:
This this infact allow in PG🤔